docs/adr/0008-freeing-objects-and-resources.md

```diff
@@ -0,0 +1,158 @@
+  which may be called before the object is destroyed, the object must deal with (a) the possibility that the object continues to be used
+  in some capacity after this close and (b) that as the object is destroyed `close()` may or may not have been called.
+* Good, because the same general solution works for when Rust is holding a foreign reference.
+* Bad, because generating "foreign sugar" means each object must opt-in to this behaviour instead of getting it for free.
+* Good, because making resource closing the user's responsibility extends to object containers as well.
+  If a library authors decides that it makes sense to define an enum variant that contains some resource-holding object, then it's the library consumer's responsibility to ensure that resource is properly closed.
+
+### [Option 2] Resource and Object as separate types
+
+* Good, because users explicitly decide on their interface behavior.
+* Good, because it means we could infer the "foreign sugar" situation and automatically implement special interfaces like `AutoCloseable`
+  in Kotlin for a Context Manager in Python.
+* Bad, because it means all objects must use the same signature for closing objects, whereas, eg, a connection to a database
+  object might prefer the method to be called `close_connection` and might even allow parameters.
+* Bad, because it implicitly defines some methods (eg, `close()`), preventing users from defining methods with those names that have different semantics.
+* Bad, because it doesn't offer a reasonable solution for when Rust is holding a foreign reference.
+* Good, because the distinction makes deciding how to handle object containers easier.
+  For example, enums can contain `Object` fields, but not `Resource` fields.
+
+### [Option 3] Make Object do both (ie, assume all objects should get close methods)
+
+* Good, because it's the easiest to implement.
+* Good, because we can easily continue to implement special traits like `AutoCloseable`.
+* Bad, for all the reasons Option 1 is bad.
+* Bad, because it's not clear how to handle object containers in all cases.  Should enums be allowed to contain objects?
+
+## Decision Outcome
+
+Chosen option: **[Option 1] UniFFI handles releasing references, closing resources is largely left up to the user**
+
+# Appendix 1 - what does this mean for our foreign bindings?
+
+The above is all quite abstract - what does this mean in practice for our language bindings:
+
+## Python
+
+* Python bindings are already hooked up to `__del__` so nothing changes here.
+* Once we have the "Foreign Sugar" opt-in mechanism in place we will generate context managers for such objects.
+
+## Swift
+
+* Existing `deinit` implementation is suitable so nothing changes here.
+
+## Kotlin
+
+* UniFFI will implement a mechanism that automically frees Rust references when the corresponding foreign object is destroyed.
```

Comment:
I think one of the last issues is to update this statement so that it's clear that UniFFI would handle this behind-the-scenes and neither the library author or consumer would need to write any code to make that happen.  Does this new wording make that clear?